brave/brave-core
# Move Ethereum Ledger integration in untrusted iframe

Resolves https://github.com/brave/brave-browser/issues/24275

Security Review: https://github.com/brave/security/issues/969

This changes isolates the js libraries required to interact with the Ethereum app on Ledger devices to an iFrame in a chrome untrusted context. This uses the same framework used in the Solana hardware as https://github.com/brave/brave-core/pull/14096, but generalizes some code such that it can be reused for both Solana and Ethereum.  For example:
* Shared logic in SolanaLedgerBridgeKeyring class was moved to a new parent class that LedgerBridgeKeyring that both SolanaLedgerBridgeKeyring and EthereumLedgerBridgeKeyring inherit from
* Solana specific logic was moved from LedgerUntrustedMessagingTransport to a new child class SolanaLedgerUntrustedMessagingTransport.  Similarly, Ethereum specific logic is now in a child class EthereumLedgerUntrustedMessagingTransport.
* Common message types were kept in ledger-messages.ts, but Solana and Ethereum specific message types were separated into their own sol-ledger-messages.ts and eth-ledger-messages.ts module.

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
The test plan is very similar to the test plan for the parallel Solana untrusted [refactor](https://github.com/brave/brave-core/pull/14096), except instead of testing Solana Ledger functions, we test the Ethereum ones.
1. Test import accounts flow (chrome://wallet)
   1. Connect ledger device and enter password
   1. Run `(await navigator.hid.getDevices()).forEach((d) => d.forget())` in the js console to forget devices which have already been granted permissions, then refresh the page
   1. Click 'Import from hardware wallet' -> 'Connect' for Ethereum, Select 'Ledger', click 'Connect'
   1. Verify 'Grant Brave access to your Ledger device.' is displayed, hit 'Authorize'
   1. You should be prompted with the permissions screen, select device, hit 'Connect'
   1. Verify you are returned to 'Connect your Ledger wallet directly to your computer' screen, hit 'Connect'
   1. Verify your accounts are listed for import for import. Verify you can load 'Load more' accounts
   1. Verify you can load different accounts by switching derivation paths
   1. Select an account and hit 'Add checked accounts', verify it is added
   1. Repeat the same flow without forgetting the device first and verify no 'Authorize' screen is displayed or permission screen prompted.
1. Test signing flows (chrome://wallet-panel) for (a) sign transaction, (b) sign personal, (c) sign typed data
   1. Connect ledger device and enter password
   1. Create a Ethereum hardware account with a test net balance if you don't have one already
   1. Navigate to chrome://inspect > Other
   1. Open wallet panel, and click 'inspect' on the chrome://inspect page under the 'chrome://wallet-panel.top-chrome' item that pops up.  This is the js console for the panel
   1. Run `(await navigator.hid.getDevices()).forEach((d) => d.forget())` to forget any device which already has permissions on the panel (need to be quick since this closes not long after the panel closes)
   1. Do one of
       1. Create a standard transaction in the panel sending funds to yourself (sign transaction)
       1. Go to OpenSea and like any NFT (I'm partial to [lil nouns](https://opensea.io/assets/ethereum/0x4b10701bfd7bfedc47d50562b76b436fbb5bdb3b/3935)) by clicking the heart (sign personal)
       1. Go to https://metamask.github.io/test-dapp/, connect wallet, and sign + verify 'Sign Typed Data V3' and 'Sign Typed Data V4'
   1. Hit sign / confirm
   1. You should be shown an 'Authorize' button on the panel that says the device is disconnect, click the 'Authorize' button
   1. You should be prompted with the permissions screen, select device, hit 'Connect'
   1. Verify you are returned to the 'Hardware wallet requires Ethereum App opened on Ledger' (it may take a second)
   1. On your ledger device, you should see the the transacition / sign request you entered, confirm.
   1. Verify the post transaction / signing stuff happens: you are shown the details, get a notification when transaction is confirmed, the 'like' is registered by OpenSea, etc
   1. Repeat all steps until all of sign transaction, sign personal, and sign typed data have been completed successfully
